### ISSUE TITLE
Cannot get class on Play

### ISSUE THREAD

**Issue by turguthakki** [godotengine/godot-cpp#511]
I have two classes defined in C++ 
```class PhysicalParameters : public Resource```
 and 
```class Vessel : public Spatial```
 Vessel class has PhysicalParameters as a property registered like 
```register_property<Vessel, PhysicalParameters *>("physicalParameters", &Vessel::physicalParameters, nullptr, GODOT_METHOD_RPC_MODE_DISABLED, GODOT_PROPERTY_USAGE_DEFAULT, GODOT_PROPERTY_HINT_RESOURCE_TYPE, "PhysicalParameters");```
 Both classes has their own gdns files. I can assign Vessel class to a spatial node and create physicalParameters as a resource in place. But when i try to run the scene engine stars and stops immediately reporting 
Cannot get class 'PhysicalParameters'.

I'm using Godot_v3.2.3-stable_win64.exe as engine.
to compile, godot-cpp along with headers must be placed in Engine/godot-cpp directory
[GDtest.zip](https://github.com/godotengine/godot-cpp/files/5960383/GDtest.zip)

**Comment by Zylann** [comment]
Does it work if you try a setter instead?
Your `PhysicalParameters` is a `Resource`, so it inherits `Reference`, which means it has to be ref-counted. It is strongly recommended to make your field a `Ref<PhysicalParameters>` instead of a `PhysicalParameters*` because the latter will not hold any reference and get freed as soon as it gets out of scope.

**Comment by turguthakki** [comment]
Tried with all three of ```
  register_property<Vessel, Ref<PhysicalParameters>>("physicalParameters", &Vessel::physicalParameters, nullptr, GODOT_METHOD_RPC_MODE_DISABLED, GODOT_PROPERTY_USAGE_DEFAULT, GODOT_PROPERTY_HINT_RESOURCE_TYPE, "PhysicalParameters");
  register_property<Vessel, PhysicalParameters *>("physicalParameters", &Vessel::setPP, &Vessel::getPP, nullptr, GODOT_METHOD_RPC_MODE_DISABLED, GODOT_PROPERTY_USAGE_DEFAULT, GODOT_PROPERTY_HINT_RESOURCE_TYPE, "PhysicalParameters");
  register_property<Vessel, Ref<PhysicalParameters>>("physicalParameters", &Vessel::setPP, &Vessel::getPP, nullptr, GODOT_METHOD_RPC_MODE_DISABLED, GODOT_PROPERTY_USAGE_DEFAULT, GODOT_PROPERTY_HINT_RESOURCE_TYPE, "PhysicalParameters");
```

none worked. Perhaps I'm missing something else?

**Comment by Zylann** [comment]
> Does it work if you try a setter instead?

Try this, and:

> It is strongly recommended to make your field a Ref<PhysicalParameters> instead

Make sure your field also is of that type, not just the registration

**Comment by Zylann** [comment]
I tried the following in my test project:

```cpp

class CustomResource : public Resource {
	GODOT_CLASS(CustomResource, Resource)
public:
	void _init() {}

	int get_int_value() {
		return _value;
	}

	static void _register_methods() {
		register_method("get_int_value", &CustomResource::get_int_value);
	}

private:
	int _value = 42;
};

class CustomNode : public Node2D {
	GODOT_CLASS(CustomNode, Node2D)
public:
	void _init() {
		_custom_resource.instance();
	}

	static void _register_methods() {
		register_property<CustomNode, Ref<CustomResource> >(
				"custom_resource", &CustomNode::_custom_resource, Ref<CustomResource>());
	}

private:
	Ref<CustomResource> _custom_resource;
};

	register_class<CustomReference>();
	register_class<CustomNode>();
```

And then this script:
```gdscript
const CustomNode = preload("res://native/custom_node.gdns")
const CustomReference = preload("res://native/custom_reference.gdns")

func _ready():
	print("BEGIN test_resource_property")
	var node = CustomNode.new()
	add_child(node)
	var res = node.custom_resource
	print(res.get_int_value())
	node.queue_free()
	print("END test_resource_property")
```

Result:
```
BEGIN test_resource_property
42
END test_resource_property
```

**Comment by turguthakki** [comment]
I changed the type in the class along with the registration.

**Comment by Zylann** [comment]
No problem occured when I tested with a script, but if you use the editor to assign properties, then maybe you are hitting this: https://github.com/godotengine/godot-cpp/issues/453
Which AFAIK has not been figured out to this day...

**Comment by turguthakki** [comment]
Could be related but this one doesn't have any reset issues. Actually you can switch between test cases and it keeps data.
Here it is. You can switch test cases using defines in Vessel.h

[GDTest.zip](https://github.com/godotengine/godot-cpp/files/5978221/GDTest.zip)

**Comment by Zylann** [comment]
Oh you are using global classes as well? These also have annoying bugs in Godot 3, and they can go badly with this issue: https://github.com/godotengine/godot-cpp/issues/430#issuecomment-752233596

Basically the C++ bindings have no way of creating a script instance with the original GDNS attached, so they build one on the fly, which can confuse static typing.

Your reproduction project is a bit confusing... first of all, just NEVER store a naked pointer to a resource like you were doing originally, that will likely not work well. It's possible to do that in rare scenarios, but not here.

You seem to have a setup of your own. What are the `.in` files? Looks like templates of some sort. Let's hope they don't contribute to the issue.

In `Vessel.h`, you inherit `Resource` but you did not include `<Resource.hpp>`, so I had to add it.

When I run the game with the Ref test case, it crashes, but there is no information about the code path taken through your library. In debug Windows builds, your SConstruct does not enable debug symbols. I added `-ZI` to include these, and I also added the `-DEBUG` option for the linker otherwise it would trigger an error.
```py
  if target == "debug":
    env.Append(CCFLAGS=["-EHsc", "-D_DEBUG", "/MDd", "-ZI"])
    env.Append(LINKFLAGS = ['-DEBUG'])
```

Now that I can debug your library, I see the crash occurs because the bindings had a fallback on `Object::_init()` to instanciate your `PhysicalParameters` class. `Obejct::_init()` is not defined (and will never be).
So the problem is you did not define `_init()` in your `PhysicalParameters` class. It is required to define it, even if empty.

Once I fixed this, your game runs without errors.

**Comment by turguthakki** [comment]
>Basically the C++ bindings have no way of creating a script instance with the original GDNS attached, so they build one on the fly, which can confuse static typing if you use GDScript.
I'll probably use GDScript for tools. But other than that C++ is the main attraction.

>You seem to have a setup of your own. What are the .in files? Looks like templates of some sort. I'll modify paths to target my own Godot and bindings.
A feeble attempt to auto generate registry of classes, properties etc. They are disabled.

> When I run the game with the Ref test case, it crashes, but there is no way to see what happened. In debug Windows builds, your SConstruct does not enable debug symbols.
I copied the SConstruct file in gdnative-demos/cpp/simple and modified it a bit. 

> So the problem is you did not define _init() in your PhysicalParameters class. It is required to define it, even if empty.
Now, only after you tell it I read the "GDNative C++ example" very carefully and found it mentioned there. 

Perhaps a "Common pitfalls" section can be placed in the documentation along with a more complex example? I heavily use nested classes to group various parameters in game entities. Such as "a characters::buffs/debuffs" or "a ships::engines". so first thing I tried was this. (They did work with C# but as I said the main attraction is C++)

**Comment by Calinou** [comment]
> Perhaps a "Common pitfalls" section can be placed in the documentation along with a more complex example? I heavily use nested classes to group various parameters in game entities. Such as "a characters::buffs/debuffs" or "a ships::engines". so first thing I tried was this. (They did work with C# but as I said the main attraction is C++)

Done: https://github.com/godotengine/godot-docs/pull/4665
